include/shm_subscriber.hpp

```diff
@@ -17,88 +17,110 @@ namespace shm_transport {
         }
-        // FIXME this segment should be locked
-        uint32_t * ptr = (uint32_t *)pshm_->get_address_from_handle(actual_msg->data);
-        M msg;
-        ros::serialization::IStream in((uint8_t *)(ptr + 2), ptr[1]);
-        ros::serialization::deserialize(in, msg);
-        // FIXME is boost::atomic rely on x86?
-        if (reinterpret_cast< boost::atomic< uint32_t > * >(ptr)->fetch_sub(1, boost::memory_order_relaxed) == 1) {
+      // FIXME this segment should be locked
+      uint32_t * ptr = (uint32_t *)pshm_->get_address_from_handle(actual_msg->data);
+      M msg;
+      ros::serialization::IStream in((uint8_t *)(ptr + 2), ptr[1]);
+      ros::serialization::deserialize(in, msg);
+      // FIXME is boost::atomic rely on x86?
+      if (reinterpret_cast< boost::atomic< uint32_t > * >(ptr)->fetch_sub(1, boost::memory_order_relaxed) == 1) {
           pshm_->deallocate(ptr);
         }
-        fp_(boost::make_shared< M >(msg));
-      }
 
-    private:
-      boost::interprocess::managed_shared_memory * pshm_;
-      std::string topic_;
-      Func fp_;
-      boost::shared_ptr< ros::Subscriber > sub_;
+      fp_(boost::make_shared< M >(msg));
+    }
+
+  private:
+    boost::interprocess::managed_shared_memory * pshm_;
+    std::string topic_;
+    Func fp_;
+    boost::shared_ptr< ros::Subscriber > sub_;
 
     friend class Topic;
     friend class Subscriber<M>;
   };
 
   template <class M>
   class Subscriber {
-    public:
-      Subscriber(const Subscriber & s) {
-        sub_ = s.sub_;
-        phlp_ = s.phlp_;
-      }
-    private:
-      Subscriber(const ros::Subscriber & sub, SubscriberCallbackHelper< M > * phlp) {
-        sub_ = boost::make_shared< ros::Subscriber >(sub);
-        phlp_ = phlp;
-        phlp_->sub_ = sub_;
-      }
+  private:
+    Subscriber(const ros::Subscriber & sub_, SubscriberCallbackHelper< M > * phlp) {
+      impl_ = boost::make_shared<Impl>();
+
+      impl_->sub = boost::make_shared< ros::Subscriber >(sub_);
+      impl_->phlp = phlp;
+      impl_->phlp->sub_ = boost::make_shared< ros::Subscriber >(sub_);
+    }
 
+    class Impl {
     public:
-      ~Subscriber() {
-        if (phlp_)
-          delete phlp_;
-        if (sub_)
-          shutdown();
+      Impl() {
       }
 
-      void shutdown() {
-        sub_->shutdown();
+      ~Impl() {
+        if (phlp) {
+            delete phlp;
+          }
+        if (sub) {
+            sub->shutdown();
+          }
       }
 
-      std::string getTopic() const {
-        return sub_->getTopic();
-      }
+      boost::shared_ptr< ros::Subscriber > sub;
+      SubscriberCallbackHelper< M > * phlp;
+    };
 
-      uint32_t getNumPublishers() const {
-        return sub_->getNumPublishers();
-      }
+    boost::shared_ptr<Impl> impl_;
+
+  public:
+    Subscriber() {
+    }
+
+    ~Subscriber() {
+    }
+
+    Subscriber(const Subscriber & s) {
+      impl_ = s.impl_;
+    }
+
```

Comment:
impl_ is a boost::shared_ptr type, so it will increase its use_count when assigned to another one.We did this to make assignment and copy consructor doing the same thing by not overloading operator =. 

Also, we only need to construct the same managed object once when doing assign or copy operation, cause they share the storage of the object.

More details please refer to the implementation of [ros::Publisher](https://github.com/ros/ros_comm/blob/lunar-devel/clients/roscpp/src/libros/publisher.cpp).